### PR TITLE
Prevent DNT checking from setting cookies

### LINKED ISSUE #1330: Cookies being set via DNT policy check requests
They should not be!

Previously:
- https://github.com/EFForg/privacybadgerfirefox-legacy/issues/220
- https://github.com/EFForg/privacybadgerfirefox-legacy/compare/0.2.2...0.2.3

### PR DESCRIPTION
**Pull request by ghostwords**
Fixes #1330.
- [x] Write test
- [x] Write fix
- [x] Get test URLs running
- [x] Document test URL logic
- [x] Enable test, verify test fails
- [x] Commit fix, verify test passes. Update: test passes in Chrome, fails in Firefox.
- [x] Confirm Firefox bug, disable test in Firefox, open ticket to track Firefox bug
- [x] Note [Channel ID](https://github.com/EFForg/privacybadger/issues/1135) connection

### CODE REVIEW

**Comment by ghostwords** [comment]
Here is a Flask version of what should be the logic behind `TEST_URL` (the deployed logic is all in Nginx):
```python
from flask import Flask, Response

app = Flask(__name__)

COOKIE = 'foo=bar; Path=/;'

@app.route('/.well-known/dnt-policy.txt')
def route_dnt_policy():
    response = Response()
    response.status = '302'
    response.headers['Set-Cookie'] = COOKIE
    response.headers['Location'] = '/cookie'
    return response

@app.route('/cookie')
def route_cookie():
    response = Response()
    response.headers['Set-Cookie'] = COOKIE
    return response

@app.route('/')
def route_index():
    return Response()

app.run('127.0.0.1', debug=True, port=443, ssl_context='adhoc')
```

**Comment by ghostwords** [comment]
Cookies should now never get set in Chrome as a result of a DNT check. Channel ID cookies do get set however, which I don't think should happen. We can track fixing this bit as part of https://github.com/EFForg/privacybadger/issues/1135.

Firefox seems to have a WebExtensions API edge case bug that lets a cookie get set from a cancelled 3xx redirect with a Set-Cookie header. Will double-check, then open a Bugzilla issue and disable the test on Firefox for now.

**Comment by ghostwords** [comment]
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1360052.

**Comment by cowlicks** [comment]
Could the issue with firefox be that `_isChromeTabInternal` is broken for Firefox? I vaguely remember a discrepancy between what the `tab.id` was for  an internal tab on firefox vs chrome.

File: tests/selenium/cookie_test.py [hunk 1]
```diff
@@ -17,6 +23,48 @@
...
+        CHECK_FINISHED_JS = "return window.DNT_CHECK_RESULT === false"
+
+        # verify that the domain itself doesn't set cookies
+        self.load_url(TEST_URL)
+        self.assertEqual(len(self.driver.get_cookies()), 0,
+            "No cookies initially")
+
+        # directly visit a DNT policy URL known to set cookies
+        self.load_url(TEST_URL + ".well-known/dnt-policy.txt")
+
```
Comment (cowlicks) [inline tests/selenium/cookie_test.py]: Should we verify the cookie here instead of after loading `TEST_URL` again below?

File: tests/selenium/cookie_test.py [hunk 2]
```diff
@@ -17,6 +23,48 @@
...
+        # verify we got a cookie
+        self.load_url(TEST_URL)
+        self.assertEqual(len(self.driver.get_cookies()), 1,
```
Comment (cowlicks) [inline tests/selenium/cookie_test.py]: This can probably stay though, to make sure it doesn't add another cookie or something

**Comment by ghostwords** [comment]
@cowlicks Not here, right, since we do get inside the `if` in Firefox. The request gets cancelled (we don't get redirected), but Set-Cookie gets incorrectly honored.

File: tests/selenium/cookie_test.py (hunk 1 above)
Comment (ghostwords) [inline tests/selenium/cookie_test.py]: Why, does it matter?
